cf-java-logging-support-log4j2/src/main/java/com/sap/hcp/cf/log4j2/layout/supppliers/BaseFieldSupplier.java

```diff
@@ -0,0 +1,38 @@
+package com.sap.hcp.cf.log4j2.layout.supppliers;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.logging.log4j.core.LogEvent;
+
+import com.sap.hcp.cf.log4j2.converter.api.Log4jContextFieldSupplier;
+import com.sap.hcp.cf.logging.common.Defaults;
+import com.sap.hcp.cf.logging.common.Fields;
+
+public class BaseFieldSupplier implements Log4jContextFieldSupplier {
+
+    @Override
+    public Map<String, Object> map(LogEvent event) {
+        Map<String, Object> fields = new HashMap<>(6);
+        fields.put(Fields.WRITTEN_AT, getIsoTs(event));
+        fields.put(Fields.WRITTEN_TS, getNanoTs(event));
+        fields.put(Fields.TYPE, LogEventUtilities.isRequestLog(event) ? Defaults.TYPE_REQUEST : Defaults.TYPE_LOG);
+        fields.put(Fields.LEVEL, String.valueOf(event.getLevel()));
+        fields.put(Fields.LOGGER, event.getLoggerName());
+        if (!LogEventUtilities.isRequestLog(event) && event.getMessage() != null) {
+            fields.put(Fields.MSG, LogEventUtilities.getFormattedMessage(event));
+        }
+        return fields;
+    }
+
+    private String getIsoTs(LogEvent event) {
+        org.apache.logging.log4j.core.time.Instant instant = event.getInstant();
+        return Instant.ofEpochSecond(instant.getEpochSecond(), instant.getNanoOfSecond()).toString();
+    }
+
+    private String getNanoTs(LogEvent event) {
+        org.apache.logging.log4j.core.time.Instant instant = event.getInstant();
+        return String.valueOf(instant.getEpochSecond() * 1_000_000_000L + instant.getNanoOfSecond());
```

Comment:
Isn't [written_at defined](https://github.com/SAP/cf-java-logging-support/blob/master/cf-java-logging-support-core/beats/app-logs/docs/fields.asciidoc#written_ts) as a long variable?
just wondering if this is for json precision issues, or actually an oversight.

Reply:
The nano timestamp is used for written_ts. This implementation was used before and is working with our logging services. So far epoch second seems to be small enough.

Reply:
Also wondering about this, because we had some precision issues with the number format of json being to small to handle nano timestamps.
We should investigate this further, as it might be a solution to the nano second rounding issues, or at least see to it being consistent with what will be ingested in elasticsearch, if we can interpret it as strings all the way into elasticsearch.